File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        prod=1
        ans=nums[0]
        for i in range(len(nums)):
            prod*=nums[i]
            ans=max(ans,max(nums[i],prod))
            if prod==0:
                prod=1
        prod=1
        for i in range(len(nums)-1,-1,-1):
            prod*=nums[i]
            ans=max(ans,max(nums[i],prod))
            if prod==0:
                prod=1
        return ans
        

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int ans=nums[0];
        int pre=1;
        int suff=1;
        for(int i=0;i<nums.length;i++){
            if(pre==0){
                pre=1;
            }
            if(suff==0){
                suff=1;
            }
            pre*=nums[i];
            suff*=nums[nums.length-i-1];
            ans=Math.max(ans,Math.max(pre,suff));
        }
        return ans;
    }
}